File: src/main/java/de/darkatra/femtojar/FemtojarExtension.java
package de.darkatra.femtojar;

import org.gradle.api.NamedDomainObjectContainer;
import org.gradle.api.Project;
import org.gradle.api.model.ObjectFactory;
import org.gradle.api.provider.Property;
import org.gradle.api.provider.ProviderFactory;
import org.gradle.api.tasks.Input;

import javax.inject.Inject;
import java.util.List;

public class FemtojarExtension {

    private final Property<Boolean> skip;
    private final Property<String> compressionMode;
    private final Property<Boolean> bundleResources;
    private final NamedDomainObjectContainer<JarEntry> jars;

    @Inject
    public FemtojarExtension(ObjectFactory objects) {
        this.skip = objects.property(Boolean.class).convention(false);
        this.compressionMode = objects.property(String.class).convention("DEFAULT");
        this.bundleResources = objects.property(Boolean.class).convention(true);
        this.jars = objects.domainObjectContainer(JarEntry.class);
    }

    public Property<Boolean> getSkip() {
        return skip;
    }

    public Property<String> getCompressionMode() {
        return compressionMode;
    }

    public Property<Boolean> getBundleResources() {
        return bundleResources;
    }

    public NamedDomainObjectContainer<JarEntry> getJars() {
        return jars;
    }
}

File: src/main/kotlin/de/darkatra/femtojar/FemtojarExtension.kt
package de.darkatra.femtojar

import me.bechberger.femtojar.rt.BundleBootstrap
import org.gradle.api.file.ProjectLayout
import org.gradle.api.file.RegularFileProperty
import org.gradle.api.model.ObjectFactory
import org.gradle.api.provider.Property
import org.gradle.api.provider.Provider
import org.gradle.api.provider.ProviderFactory
import javax.inject.Inject

open class FemtojarExtension @Inject constructor(
    objects: ObjectFactory,
    private val layout: ProjectLayout,
    providers: ProviderFactory,
) {

    internal val inputFile: RegularFileProperty = objects.fileProperty()
    internal val outputFile: RegularFileProperty = objects.fileProperty()

    var `in`: String? = null
        set(value) {
            field = value
            if (value == null) {
                inputFile.unset()
            } else {
                inputFile.set(layout.projectDirectory.file(value))
            }
        }

    var out: String? = null
        set(value) {
            field = value
            if (value == null) {
                outputFile.unset()
            } else {
                outputFile.set(layout.projectDirectory.file(value))
            }
        }

    val skip: Property<Boolean> = objects.property(Boolean::class.java).convention(false)
    val compressionMode: Property<String> = objects.property(String::class.java).convention("DEFAULT")
    val bundleResources: Property<Boolean> = objects.property(Boolean::class.java).convention(true)

    val originalMainClass: Provider<String> = providers.of(MainClassValueSource::class.java) { spec ->
        spec.parameters.inputFile.set(inputFile)
    }
    val bootstrapMainClass: Provider<String> = providers.provider {
        BundleBootstrap::class.qualifiedName!!
    }
}
